<a href="https://colab.research.google.com/github/DesiPilla/DSPS_dPilla/blob/master/HW5/BuildingsEnergyOwners.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [47]:
# -- link google drive
from google.colab import drive
drive.mount("/content/drive")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [0]:
# -- useful imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as pl
import scipy as sp
from scipy import stats

# data retrival
### LL84
get the Local Law 84 (ll84) data from the URL: https://data.cityofnewyork.us/api/views/8u86-bviy/rows.csv and read it in with Pandas

### Pluto
Download the pluto csv data and store it in your google drive. 

You can find the data here 
https://www1.nyc.gov/site/planning/data-maps/open-data/dwn-pluto-mappluto.page

IMPORTANT: store it in the path *drive/My Drive/dsps/ with the name pluto_18v2_1.csv*

This path and name have to be exact. This way the grader and I will be able to run your notebook without changing anything as long as we also have stored the data in path drive/My Drive/dsps/pluto_18v2_1.csv. This is part of the _reproducibility_ requirement.

In [0]:
ll84 = pd.read_csv('https://data.cityofnewyork.us/api/views/8u86-bviy/rows.csv')

In [50]:
mkdir "drive/My Drive/dsps"

mkdir: cannot create directory ‘drive/My Drive/dsps’: No such file or directory


In [51]:
cd "drive/My Drive/dsps"

[Errno 2] No such file or directory: 'drive/My Drive/dsps'
/content/drive/My Drive/dsps/drive/My Drive/dsps


In [52]:
ls

pluto_18v2_1.csv


# data fusion

### select data
select the data for buildings _in the borough of Manhattan_ only from the pluto file (you can do it broadcasting!)

### merge
merge the pluto and ll84 dataframes. you will have to choose an appropriate column that both datasets share fo do so. The column may not have the same nbame, but should have the same _variable_.

In [81]:
pluto = pd.read_csv("pluto_18v2_1.csv")
pluto.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (16,17,18,20,22,77) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


borough   block    lot     cd  ...  basempdate  masdate  polidate  edesigdate
0      BX  5641.0  670.0  210.0  ...  12/21/2018      NaN       NaN  12/20/2018
1      QN  3935.0    3.0  407.0  ...  12/21/2018      NaN       NaN  12/20/2018
2      BX  4205.0   41.0  211.0  ...  12/21/2018      NaN       NaN  12/20/2018
3      BX  2402.0   14.0  201.0  ...  12/21/2018      NaN       NaN  12/20/2018
4      SI    63.0   83.0  501.0  ...  12/21/2018      NaN       NaN  12/20/2018

[5 rows x 96 columns]

In [82]:
pluto = pluto[pluto["borough"] == "MN"]
pluto.shape

(42897, 96)

In [83]:
ll84.rename(columns={"Address 1 (self-reported)": "address"}, inplace = True)

joined = pd.merge(pluto, ll84, on='address', how='left')
joined.shape

(42897, 155)

# data exploration and cleaning

### extact columns
extract the columns listed in the file "columns.txt" from the dataset.

convert the resulting dataset to numerical, converting non-numerical values to NaN. drop rows that contain invalid values and save the resulting dataframe (_reducted_ dataframe)

In [127]:
columns = pd.read_table("https://raw.githubusercontent.com/fedhere/DSPS/master/HW5/columns.txt", header = None)
columns = columns[0].values.tolist()
print(columns)


data = joined[columns]
data = data.convert_objects(convert_numeric = True)
reduced = data.dropna(thresh = 22)
reduced.shape

['DOF Gross Floor Area', 'Year Built', 'Number of Buildings - Self-reported', 'Occupancy', 'ENERGY STAR Score', 'Site EUI (kBtu/ft²)', 'Weather Normalized Site EUI (kBtu/ft²)', 'Weather Normalized Site Electricity Intensity (kWh/ft²)', 'Weather Normalized Site Natural Gas Intensity (therms/ft²)', 'Weather Normalized Source EUI (kBtu/ft²)', 'Electricity Use - Grid Purchase (kBtu)', 'Total GHG Emissions (Metric Tons CO2e)', 'Property GFA - Self-Reported (ft²)', 'Water Use (All Water Sources) (kgal)', 'Water Intensity (All Water Sources) (gal/ft²)', 'Source EUI (kBtu/ft²)', 'bldgarea', 'numbldgs', 'numfloors', 'unitsres', 'unitstotal', 'yearalter1']


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:1: FutureWarning: read_table is deprecated, use read_csv instead, passing sep='\t'.
  """Entry point for launching an IPython kernel.
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:7: FutureWarning: convert_objects is deprecated.  To re-infer data dtypes for object columns, use DataFrame.infer_objects()
For all other conversions use the data-type specific converters pd.to_datetime, pd.to_timedelta and pd.to_numeric.
  import sys


(14, 22)

In [129]:
reduced.head()

DOF Gross Floor Area  Year Built  ...  unitstotal  yearalter1
1504                91212.0      1927.0  ...        56.0         0.0
2912                63600.0      1920.0  ...        30.0         0.0
4008                55665.0      1900.0  ...        43.0         0.0
4769               137500.0      1990.0  ...       164.0         0.0
11021               85194.0      1912.0  ...       104.0      2006.0

[5 rows x 22 columns]

### Assess impact of data wrangling
Describe the  datasets resulting from this data cleaning: 

- how many observations have you lost? 

- did any of the features (=columns) have inconsistent distribution between the full and reducted dataset? 

You can approach this question in a number of ways, and you have done this in the  homewors already (with a KS test, and by just looking at uncertainties are two examples of ways to do ut that you worked out in HW3 and HW4, the Z test might work if you can convince yourself that the full sample is the "population" - up to you! a quick google search would probably describe many more ways). Choose _2_ ways, describe what each does and why it is appropriate. Compare the results (indeed I did it in 4 ways and did not always get consistent results! each of these meathods measures a slightly different thing).



**My answer:**

We started with 42,897 observations when we first joined the data frames. After eliminating all NaN / non-numeric answers, we were left with only 14 observations.

In [147]:
reducedStats = {}
for col in columns:
  reducedStats[col] = [np.average(reduced[col]), np.std(reduced[col])]
reducedStats = pd.DataFrame.from_dict(reducedStats, orient = "index", columns = ["mean", "stdev"])
reducedStats

mean         stdev
DOF Gross Floor Area                                2.797568e+05  4.383691e+05
Year Built                                          1.941786e+03  3.475872e+01
Number of Buildings - Self-reported                 1.000000e+00  0.000000e+00
Occupancy                                           9.035714e+01  2.580589e+01
ENERGY STAR Score                                   5.614286e+01  2.861033e+01
Site EUI (kBtu/ft²)                                 7.977143e+01  2.365854e+01
Weather Normalized Site EUI (kBtu/ft²)              8.233571e+01  2.488791e+01
Weather Normalized Site Electricity Intensity (...  1.032143e+01  5.976045e+00
Weather Normalized Site Natural Gas Intensity (...  2.142857e-01  2.924213e-01
Weather Normalized Source EUI (kBtu/ft²)            1.600000e+02  5.225596e+01
Electricity Use - Grid Purchase (kBtu)              1.698236e+07  3.797439e+07
Total GHG Emissions (Metric Tons CO2e)              2.132771e+03  4.204557e+03
Property GFA - Self-Reported (ft²)                  2.791238e+05  4.276480e+05
Water Use (All Water Sources) (kgal)                9.175421e+03  1.037736e+04
Water Intensity (All Water Sources) (gal/ft²)       5.381000e+01  6.401782e+01
Source EUI (kBtu/ft²)                               1.580857e+02  5.265757e+01
bldgarea                                            2.771178e+05  4.394129e+05
numbldgs                                            1.000000e+00  0.000000e+00
numfloors                                           1.592857e+01  1.047762e+01
unitsres                                            5.214286e+01  6.820123e+01
unitstotal                                          6.742857e+01  6.373126e+01
yearalter1                                          9.975000e+02  9.975379e+02

In [141]:
np.std

<function numpy.std>

### plot WNEC
Regardless of what you found above, plot the distribution of the _log base 10_ of the weather-normalized site energy consumption (WNEC) for all values larger than 0.1 (such that log10(min) is -1) for the _entire dataset_ and for the _reducted one_: do you think inference on this feature would be affected by this way of dealing with missing data affects the result?



# 667 and extyra credit
plot a "scatter matrix" of the selected columns. Identify 4 promising "scaling laws" and plot them separately (a scatter plot of column1 vs column2). Does it make sense that there should be a scaling law between the 2 variables? 

which features should be strongly correlated or identical but are not? discuss

([this](https://stackoverflow.com/questions/26975089/making-the-labels-of-the-scatterplot-vertical-and-horizontal-in-pandas) stackoverflow may come in handy)

## all students
At last, from the original dataset remove only the columns that are NaN in the column "Weather Normalized Site EUI (kBtu/ft²)"


In [0]:
# -- drop buildings for which WN Site EUI is NaN
wnec = "Weather Normalized Site EUI (kBtu/ft²)"
...
ll84pl.shape

## Analysis: Owner investigation 
### Find owner of the single most energy consuming building

**Find owner of the single most energy consuming building**, as measured by the "Site Weather Normalized EUI per unit area ". 

**Find all the buildings that they own.**


show the weather-normalized site energy consumption and the owner name for the top 30 consuming buildings in Manhattan

In [0]:
print("Owner of the hisghest consuming building:...", 
      "They own ... buildings")

### Find average building consumption by owner 
For all owners that own more than one building, find the average consumption for buildings owned by the same owner. Your dataframe should have 2 columns called _Weather Normalized Site EUI (kBtu/ft²)_	and _counts_

**Who owns the most energy consuming buildings on average?**

**Who owns the most buildings**

# 667 and EC

Universities are big realtors in NYC. The may own by different names. Find all builgings owned by NYU and COLUMBIA. replace the ownename with NYU and COLUMBIA respectively and re-aggregate the data based on these names and again find the owner of the buildings that pon average consume the most, of the highest number of buildings, and that owns the most sq footage 